lib/deimos/config/configurable.rb

```diff
@@ -0,0 +1,258 @@
+  #       class_name 'MyProducer'
+  #       topic 'MyTopic'
+  #     end
+  #   end
+  # - Allows to call `configure` multiple times without crashing.
+  # - Allows to lazy-set default values by passing a proc as a default:
+  #  Deimos.configure do |config|
+  #   setting :my_val, default_proc: proc { MyDefault.calculated_value }
+  #  end
+  # - Support for setting up and automatically calling deprecated configurations.
+  # - Support for configuration callbacks.
+  module Configurable
+    extend ActiveSupport::Concern
+
+    ConfigSetting = Struct.new(:value, :default_value, :default_proc, :deprecation) do
+
+      # Reset value back to default.
+      def reset!
+        if self.value.is_a?(ConfigStruct)
+          self.value.reset!
+        elsif self.default_proc
+          self.value = self.default_proc.call
+        else
+          self.value = self.default_value
+        end
+      end
+
+      # :nodoc:
+      def clone
+        setting = ConfigSetting.new(self.value, self.default_value,
+                                    self.default_proc, self.deprecation)
+        setting.reset!
+        setting
+      end
+
+    end
+
+    # Class that defines and keeps the configuration values.
+    class ConfigStruct
+      include ActiveSupport::Callbacks
+
+      define_callbacks :configure
+
+      # @param name [String]
+      def initialize(name)
+        @name = name
+        @settings = {}
+        @setting_objects = {}
+        @setting_templates = {}
+      end
+
+      # Reset config back to default values.
+      def reset!
+        @setting_objects = @setting_templates.map { |k, _| [k, []] }.to_h
+        @settings.values.each(&:reset!)
+      end
+
+      # Mark a configuration as deprecated and replaced with the new config.
+      # @param old_config [String]
+      # @param new_config [String]
+      def deprecate(old_config, new_config)
+        @settings[old_config.to_sym] ||= ConfigSetting.new
+        @settings[old_config.to_sym].deprecation = new_config
+      end
+
+      # :nodoc:
+      def inspect
+        "#{@name}: #{@settings.inspect} #{@setting_objects.inspect}"
+      end
+
+      # @return [Hash]
+      def to_h
+        @settings.map { |k, v| [k, v.value] }.to_h
+      end
+
+      # :nodoc:
+      def clone
+        new_config = super
+        new_config.setting_objects = new_config.setting_objects.clone
+        new_config.settings = new_config.settings.map { |k, v| [k, v.clone] }.to_h
+        new_config
+      end
+
+      # Define a setting template for an array of objects via a block:
+      #   setting_object :producer do
+      #     setting :topic
+      #     setting :class_name
+      #   end
+      # This will create the `producer` method to define these values as well
+      # as the `producer_objects` method to retrieve them.
+      # @param name [Symbol]
+      def setting_object(name, &block)
+        new_config = ConfigStruct.new("#{@name}.#{name}")
+        @setting_objects[name] = []
```

Comment:
May be good to add a warning if name is redefined

Reply:
I think this might get noisy. Extending settings is supported fully by the framework so I don't see this as a warning.